### ISSUE TITLE
spare disks don't seem to work properly

### ISSUE THREAD

**Issue by jimsalterjrs** [zfsonlinux/zfs#4677]
Created a three-disk RAIDZ1 vdev.  Added two disks as spares.  Set autoreplace=yes.  

Faulted one disk by 'physically' removing it: hotspare did not deploy automatically, even after pool set itself DEGRADED.

````
NAME        STATE     READ WRITE CKSUM
test        DEGRADED     0     0     0
  raidz1-0  DEGRADED     0     0     0
    nbd0    UNAVAIL      0    70     0  corrupted data
    nbd1    ONLINE       0     0     0
    nbd2    ONLINE       0     0     0
spares
  nbd3      AVAIL   
  nbd4      AVAIL   
````

Gave it some time, spare never deployed.  Manually deployed spare, got an odd 'in use' error message which I assume is itself evidence of a bug:

````
root@banshee:/tmp# zpool replace test nbd0 nbd3
cannot open '/dev/nbd3': Device or resource busy
````

Error message aside, this did mark the requested spare disk as in use:

````
root@banshee:/tmp# zpool status test | grep -A10 NAME
NAME         STATE     READ WRITE CKSUM
test         DEGRADED     0     0     0
  raidz1-0   DEGRADED     0     0     0
    spare-0  UNAVAIL      0     0     0
      nbd0   UNAVAIL      0    70     0  corrupted data
      nbd3   ONLINE       0     0     0
    nbd1     ONLINE       0     0     0
    nbd2     ONLINE       0     0     0
spares
  nbd3       INUSE     currently in use
  nbd4       AVAIL   
````

But spare-0 shows as UNAVAIL - with no message about resilvering at all, much less a progress report for percentage complete or speed of resilvering.

Given enough time, it eventually does resilver, after which spare-0 shows DEGRADED instead of UNAVAIL.  There never is any message at all about resilvering, though - in process *or* complete.  I consider this a pretty large bug, given that multi-terabyte hotspares could easily take hours or even days to resilver.

**Comment by jimsalterjrs** [comment]
After a little more experimentation, it's worse than I thought: `spare-0` NEVER automatically resilvers and shifts from UNAVAIL to DEGRADED.  It only becomes DEGRADED after I manually trigger a scrub of the pool.

````
root@banshee:/tmp# zpool replace test nbd0 nbd3
cannot open '/dev/nbd3': Device or resource busy
root@banshee:/tmp# zpool status test
  pool: test
 state: DEGRADED
status: One or more devices could not be used because the label is missing or
	invalid.  Sufficient replicas exist for the pool to continue
	functioning in a degraded state.
action: Replace the device using 'zpool replace'.
   see: http://zfsonlinux.org/msg/ZFS-8000-4J
  scan: resilvered 32.3M in 0h0m with 0 errors on Fri May 20 18:55:05 2016
config:

	NAME         STATE     READ WRITE CKSUM
	test         DEGRADED     0     0     0
	  raidz1-0   DEGRADED     0     0     0
	    spare-0  UNAVAIL      0     0     0
	      nbd0   UNAVAIL      0     0     0  corrupted data
	      nbd3   ONLINE       0     0     0
````

I gave the vdev a good 2 minutes or so to resilver onto nbd3 by itself - it never did.  Then I scrubbed the pool:

````
root@banshee:/tmp# zpool scrub test
root@banshee:/tmp# zpool status test
  pool: test
 state: DEGRADED
status: One or more devices could not be used because the label is missing or
	invalid.  Sufficient replicas exist for the pool to continue
	functioning in a degraded state.
action: Replace the device using 'zpool replace'.
   see: http://zfsonlinux.org/msg/ZFS-8000-4J
  scan: scrub repaired 0 in 0h0m with 0 errors on Fri May 20 18:56:27 2016
config:

	NAME         STATE     READ WRITE CKSUM
	test         DEGRADED     0     0     0
	  raidz1-0   DEGRADED     0     0     0
	    spare-0  DEGRADED     0     0     0
	      nbd0   UNAVAIL      0     0     0  corrupted data
	      nbd3   ONLINE       0     0     0
````

Scrubbing the pool triggered a resilver, effectively at least, and spare-0 THEN went from UNAVAIL to DEGRADED.

TL;DR spares are basically entirely useless right now - they won't autodeploy, and they won't automatically resilver even WHEN deployed.  In order to make use of one, you have to both manually deploy it with a replace command (which triggers an error message that it shouldn't), then in order to resilver it and actually get any use out of it, you have to scrub the pool (potentially generating a lot more IO than necessary during a vulnerable window).

AFAICT, right now you're considerably better off leaving "spare" disks physically in a server *but completely unassigned*, and manually replacing them in after a failure, than you are actually using the hotspare feature right now.

Am I missing something?

**Comment by FransUrbo** [comment]
> Created a three-disk RAIDZ1 vdev. Added two disks as spares. Set autoreplace=yes.
> 
You need to configure ZED (ZFS Event Daemon) to do this.

**Comment by jimsalterjrs** [comment]
> You need to configure ZED

Is that also the issue with the spare, once deployed, not automatically resilvering?

**Comment by FransUrbo** [comment]
> You need to configure ZED
> 
> Is that also the issue with the spare, once deployed, not automatically resilvering?
> 
No, it should do that.

**Comment by jimsalterjrs** [comment]
Is there a place I should be looking for info on configuring ZED?  All I'm finding from googling are two year old blog posts about it automatically being installed with OpenZFS, only needing configuration if you want it to send emails, and it being buggy about sending the emails.

**Comment by FransUrbo** [comment]
> Is there a place I should be looking for info on configuring ZED? All I'm finding from googling are two year old blog posts about it automatically being installed with OpenZFS, only needing configuration if you want it to send emails, and it being buggy about sending the emails.
> 

It's supposed to be in /etc/zfs/zed.d. But also check zed(8).

**Comment by jimsalterjrs** [comment]
Additionally, while recreating this issue and trying to capture and annotate the contents of `zpool events` at various points during, something went completely haywire in the ZFS stack after doing `qemu-nbd -d /dev/nbd1` (to "yank" a drive from the running raidz1) - no `zpool status` commands will complete any longer, even of different pools on the same machine, and the pool cannot be exported, destroyed, or otherwise poked or prodded at in any way to get things back under control either.  Several hours later, my system load is up to 140+ due to failing commands run from cron (automatically taking and rotating snapshots), and I'm going to eventually have to reboot the system to get thing working properly again.

This all seems like a pretty drastically bad response to a relatively common failure event (a drive dropping off a bus unexpectedly).

**Comment by kpande** [comment]
closing as duplicate of #2449
